### DISCUSSION TITLE
change width of a map

### DISCUSSION THREAD

**Discussion by mhechthz** [freeplane/freeplane#520] [General]
I have a large map that has not so much subnodes. Zooming this out leads to such a view:

![image](https://user-images.githubusercontent.com/37075144/182089653-5b021dea-1c44-4a2a-a201-9b59d5d6bdd7.png)

This view is much too narrow (horizontally) in my opinion. What I liked to see is something like this

![image](https://user-images.githubusercontent.com/37075144/182089986-506c585d-acce-4b93-8d00-acf246e8916e.png)

So the question would be: is there a chance to modify the horizontal expansion of the map. The proposed solution [https://github.com/freeplane/docs/issues/64#issuecomment-1200401057](url) does not work because I don't want to see the outline but the 'normal' map or to have the map more squared than vertical.

**Comment by euu2021** [comment]
There is no way to automatically change the horizontal gap, as it is possible with child gap.

To make things more horizontal, I usually increase the Maximum and Minimum Node Width of the nodes.

But, I see that you use a lot of images, so you may need to position at least the high level nodes manually. Use the oval handles to move the nodes.

**Reply by euu2021** [reply]
Increasing the horizontal margin is another solution:

From this:
![image](https://user-images.githubusercontent.com/77707706/182191114-4aa82813-a0fb-435d-bad4-3d13cb50f79a.png)


To this:
![image](https://user-images.githubusercontent.com/77707706/182191085-c0a9cce6-b8ef-4279-b16e-cb35430ea585.png)

**Comment by euu2021** [comment]
I found the method setHorizontalShift in the API:

![image](https://user-images.githubusercontent.com/77707706/182195755-a5efbc14-0163-4b93-90d6-bd9e86573772.png)

But, I don't know how to make it work and I'm not finding examples. I tried things like:
`node.horizontalShift.setHorizontalShift(10 pt)`

**Reply by macmarrum** [reply]
>But, I don't know how to make it work and I'm not finding examples. I tried things like:
`node.horizontalShift.setHorizontalShift(10 pt)`

@eu2021 a good idea to use a script to increase horizontal gap between nodes 👍 
You can use one of these:
* `node.setHorizontalShift("10 pt")` using regular Java syntax
* `node.horizontalShift = '10 pt'` using Groovy shorthand syntax

This will work on all selected nodes:
* `c.selecteds.each { it.horizontalShift = '1 cm' }`

**Reply by euu2021** [reply]
Thanks, macmarrun. I need to get the hang of how to write those expressions. I always forget something. That's why I think having working examples is helpful.

---
So, @mhechthz, another solution is to use a script. This will allow changing the actual horizontal gap:

![image](https://user-images.githubusercontent.com/77707706/182211271-6b008b22-093b-4bdb-9deb-df2d15b28bf5.png)

Use this script to apply a gap of 10 cm to all nodes (change the size according to your needs):
`c.findAll().each { it.horizontalShift = '10 cm' }`

For instructions on how to use scripts, please see [this page](https://docs.freeplane.org/#/scripting/Scripting).

I suggest making the script be applied on every Freeplane restart, so you don't need to worry about reapplying it manually when you create more nodes. See instructions in [this page](https://docs.freeplane.org/#/attic/old-mediawiki-content/Init_scripts).

**Reply by macmarrum** [reply]
> I suggest making the script be applied on every Freeplane restart, so you don't need to worry about reapplying it manually when you create more nodes. See instructions in [this page](https://docs.freeplane.org/#/attic/old-mediawiki-content/Init_scripts).

I'm not sure [Init_scripts](https://docs.freeplane.org/#/attic/old-mediawiki-content/Init_scripts) can be used to achieve it. @eu2021 I would be interested to see the script, if you have something like that working.

But there is another way to have it auto-applied per mindmap. See [New auto-colouring of nodes](https://github.com/freeplane/freeplane/discussions/424#discussioncomment-2917418) and [let leaf nodes have different colors automatically](https://github.com/freeplane/freeplane/discussions/514#discussioncomment-3255433)

**Reply by euu2021** [reply]
I put a .groovy file in the init folder with just this line, and it worked:

`c.findAll().each { it.horizontalShift = '10 cm' }`

FP started, and a split of a second later the nodes had their horizontal gaps shifted.

> But there is another way to have it auto-applied per mindmap. 

This seems interesting. I will try to understand what is happening there.

**Reply by euu2021** [reply]
> I suggest making the script be applied on every Freeplane restart, so you don't need to worry about reapplying it manually when you create more nodes.

I think you thought I was suggesting something like listeners here, but what I meant is that: new nodes won't get autoarranged, but you can leave them like that, and on the next restart they will be rearranged.

**Reply by macmarrum** [reply]
> I put a .groovy file in the init folder with just this line, and it worked

Oh, I see. I guess it worked on a single mindmap that was open automatically during FP start.
I would probably find it less work to just run the script directly on the mindmap, using `Tools->Edit script…`

**Reply by euu2021** [reply]
> Oh, I see. I guess it worked on a single mindmap that was open automatically during FP start.

Yes, so it's kind of limited. It's not very useful if the user uses several mindmaps. 

> I would probably find it less work to just run the script directly on the mindmap, using Tools->Edit script…

I did not get this part, but I would like to understand. So in this situation:

1. -User has a map with several nodes with normal gap
2. -User applies script, making all nodes get expanded gap
3. -User creates new nodes, that have normal gap

How would you use `Tools->Edit script…`? To run the script directly on the mindmap isn't it easier to use `Tools->Scripts->ScriptX`?

**Reply by macmarrum** [reply]
> I did not get this part, but I would like to understand. So in this situation:

After step 3, I'd execute step 2 again.
To be more precise, I would save the script on the root node and execute it from there in step 2 and 4.
I mean, to me it's easier than close FP and launch it again so that an init script gets run.

> How would you use `Tools->Edit script…?` To run the script directly on the mindmap isn't it easier to use `Tools->Scripts->ScriptX`?

For a single mindmap, I often find it easier to use `Tools->Edit script…`, because no restart of FP is required and the script travels with the map.
If I want to run a script on any mindmap, I use `Tools->Scripts->ScriptX`. The script is then part of this FP instance, so when I open the map in another FP instance, I need to remember to also bring the script along.

**Reply by euu2021** [reply]
> I mean, to me it's easier than close FP and launch it again so that an init script gets run.

Oh, I see, you thought I was excluding the use of the script beside the init script. But, I was suggesting something more like that:

```
1-User has a map with several nodes with normal gap
2-User applies script, making all nodes get expanded gap
3-User creates new nodes, that have normal gap
4-User does
	if (user gets bothered with the normal gap in the new nodes and wants everything with expanded gap immediatly)
		{go back to step 2}
	else
		{keep working, because, on FP restart, nodes gap will be adjusted automatically}
```

> For a single mindmap, I often find it easier to use Tools->Edit script…, because no restart of FP is required and the script travels with the map.

But, then it's possible to assign hotkeys to scripts? Or do you need to navigate to another node every time?

**Reply by euu2021** [reply]
Ok, so even without understanding, I went ahead and created the conditional style rule (script filter): 

```
 if (!node.isRoot()) return false
//
c.findAll().each { it.horizontalShift = '5 cm' }

return true
```

And it works. (@mhechthz , that's a good solution)

Let me try to guess what happens here:

- with the script filter option, the user gives a description of the condition that he wants to be found for it to be considered a match and the conditional style applied. In the case above, the condition is: node is root.
- besides that, you can insert commands to be executed, and they can be completely unrelated to the search of conditions. 
- FP will keep looping on that search all the time. 

If that's the case, then I have 2 questions:

1. What is the trigger for each search loop? Is it any map change?
2. Why FP freezes if I put as a script filter simply: `c.findAll().each { it.horizontalShift = '5 cm' }` ? In other words, I don't understand the logic of using  if ... return false ... return true

**Comment by macmarrum** [comment]
> But, then it's possible to assign hotkeys to scripts? Or do you need to navigate to another node every time?

One needs to navigate to the node with the script and execute it.

**Comment by macmarrum** [comment]
> 1. What is the trigger for each search loop? Is it any map change?

I assume that FP fires the rules/scripts/conditions on each node change. But the best way to find out how it works is to study the source code of this functionality, which I have not done so far.

> 2. Why FP freezes if I put as a script filter simply: c.findAll().each { it.horizontalShift = '5 cm' } ? In other words, I don't understand the logic of using if ... return false ... return true

It's probably because the code `c.findAll().each { it.horizontalShift = '5 cm' }` returns a non-empty list (a truthy value) for any node. And the code itself causes changes which trigger FP Map Conditional Rules to run, which in turn trigger the code to run – an endless loop.

In fact, the code `if (!node.isRoot()) return false; c.findAll().each { it.horizontalShift = '5 cm' }; return true` is only slightly better. It also causes an endless loop.
A better approach is to use FP internal code to change horizontal shift, which avoids triggering node-changed events, i.e.

```groovy
if (!node.isRoot()) return false

import org.freeplane.api.LengthUnit
import org.freeplane.api.Quantity
import org.freeplane.features.nodelocation.LocationModel

def horizontalShift = '1 cm'

def hGap = Quantity.fromString(horizontalShift, LengthUnit.px)
node.mindMap.root.findAll().drop(1).each {
    LocationModel.getModel(it.getDelegate()).setHGap(hGap)
}

return true
```

**Reply by macmarrum** [reply]
Here's how I play with the execution of a Script Filter to better understand it.

I know that a Map Conditional Style assigns a style to each node that matches a defined condition.
Script Filter is a script that defines a condition, i.e. returns a _truthy_ or a _falsy_ value _(https://docs.groovy-lang.org/next/html/documentation/#the-groovy-truth)_.

To better understand the mechanics, I create a Map Conditional rule that applies the style `Important` to a node, on the condition that the node's text is '!!!', using a Script Filter.
I can test it by creating a node with the text '!!!', and verify that it gets the style `Important` applied.
As a next step, I split the logic so that there are two separate statements, one that returns true if the condition is met, another that check for the condition not to be met and returns false.
Next, I put the condition-not-met statement at the top, and the condition-met statement at the bottom.
And between the two, I put `c.statusInfo = 'Hello from a Script Filter @ ' + node.id`

Then I change the text of the node '!!!' to meet the condition and not to. I observe the status line and the applied style.

**Reply by euu2021** [reply]
Ok, I followed your suggestion and made some tests and I think I understood how the Script Filter works. I will open another thread to avoid offtracking this one.

---
Changing the FP internal code seems very interesting, and I need to make some tests in order to understand how it works. 

Now, maybe this is another solution:
```
if (node.getHorizontalShift() == '10 cm') 
    return false
		
else
    node.horizontalShift = '10 cm' 
    return false
```
I think it avoids loops, right? (I will explain the second `return false` on the other thread)

**Reply by euu2021** [reply]
(about my script)
actually, it wasn't so simple. I noticed that there was still some loop going on, because there was CPU load. 

So, I measured the shifts and noticed that the shift changes after you set it. If I set 300, it turns into 400; if I set 400 it turns into 533. So, I solved this using >= instead of =. This will be a problem when trying to decrease the shift, after having increased it, though.

Second thing: the getHorizontalShift() works with integers, and I was using String for comparison.

So, the script that is working without a loop (but is not easy to decrease the shift):

```
if (node.getHorizontalShift() >= 300) 
    return false
		
else
    node.horizontalShift = 300
    return false
```

---
(about your script)

It's throwing me an error as soon as I create another node besides the root:
![image](https://user-images.githubusercontent.com/77707706/182454872-3736c01d-85c1-4ff8-b25b-995a3e0ddd2e.png)

**Reply by macmarrum** [reply]
> I noticed that there was still some loop going on, because there was CPU load
[...]
So, the script that is working without a loop (but is not easy to decrease the shift)

On my computer it still goes to a loop, although it stops after several iterations.
The way I understand it, the instruction `node.horizontalShift = ...` triggers a NodeChange event, which triggers execution of the Map Conditional Rule, which triggers the instruction, etc.

> It's throwing me an error as soon as I create another node besides the root:

A good catch. I made a mistake. Change `LocationModel.getModel` to `LocationModel.createLocationModel`

**Reply by euu2021** [reply]
> On my computer it still goes to a loop, although it stops after several iterations. The way I understand it, the instruction `node.horizontalShift = ...` triggers a NodeChange event, which triggers execution of the Map Conditional Rule, which triggers the instruction, etc.

Here, the CPU load stops in a split of a second after the interaction. I don't know how to see the iterations, but I think there should be only one loop (which is expected), and only if the shift is bigger than 300:
 
So, after interaction (trigger), FP checks only the changed node, and follow the steps on it, according to the scenario found:

First scenario (shift lower than 300)

1. Notice shift lower than 300 -> stop. No triggers created, no loop.

 
Second scenario (shift bigger than 300) 

1. Notice the shift than 300 -> change it to 300. Node changed, so trigger created, start again.
2. Check again the same node (and only it), and find the first scenario, i.e.: Notice shift lower than 300 -> stop. No triggers created, no loop.


> A good catch. I made a mistake. Change `LocationModel.getModel` to `LocationModel.createLocationModel`

It's seems to be perfect, now. New form:

```
if (!node.isRoot()) return false

import org.freeplane.api.LengthUnit
import org.freeplane.api.Quantity
import org.freeplane.features.nodelocation.LocationModel

def horizontalShift = '1 cm'

def hGap = Quantity.fromString(horizontalShift, LengthUnit.px)
node.mindMap.root.findAll().drop(1).each {
    LocationModel.createLocationModel(it.getDelegate()).setHGap(hGap)
}

return true
```

So, I think we can conclude that this is currently the most convenient way to change the horizontal gap.

**Reply by macmarrum** [reply]
> I don't know how to see the iterations, but I think there should be only one loop (which is expected), and only if the shift is bigger than 300

Here's how you can observe the effects of such a solution

1. Run you FP using the `freeplane.bat` file, so that you can see the Command Terminal
2. Use the following Script Filter – it prints information when executed. On my computer I can see many executions.

```
if (node.root) return true

import org.freeplane.api.LengthUnit
import org.freeplane.api.Quantity
import org.freeplane.features.nodelocation.LocationModel

def hs = '60'
def now = format(new Date(), 'HH:mm:ss.S')
println("$now $hs ${node.id} ${node.text}")

def horizontalShift = Quantity.fromString(hs, LengthUnit.px)
LocationModel.getModel(node.getDelegate()).setHGap(horizontalShift)

return false
```

I am looking into having a NodeChangeListener instead of Script Filter for this purpose.

**Reply by euu2021** [reply]
Go figure...

At least those extra iterations don't bite too much of the CPU power.

**Reply by macmarrum** [reply]
Here's a NodeChangeListener that contains `setHorizontalShift` – https://github.com/macmarrum/freeplane-scripting/blob/main/src/scripts/macmarrumNCL.groovy

Until #523 is merged, change `def listeners = node.mindMap.listeners` to `def listeners = []`.

**Reply by euu2021** [reply]
I still don't get why do you say NodeChangeListeners are the proper approach for this situation. Having to execute them on every restart seems to be a relevant disadvantage. On the other hand, I see that NodeChangeListener are the only possible approach when the user wants to limit the kinds of changes to be observed, like in the V2 of [this script](https://sourceforge.net/p/freeplane/discussion/758437/thread/e9200db6c3/).

---
I tried to create exceptions to the Script Filter by placing a condition with a Stop above it (which is something users may want to do on that and other Script Filters) and it works for my script:
![image](https://user-images.githubusercontent.com/77707706/182592176-69bcaba2-1877-4877-b839-2bf02e794e5f.png)

But, it doesn't work for your script. It works if I set `always` as the condition, but it doesn't work (i.e., hshift gets changed anyway) if I set a Style or Core Text as condition. Is it because you use the approach of changing the hshift globally? If so, I don't understand why the `always` condition works.

**Comment by dpolivaev** [comment]
I want to give an explicit warning:
neither formulas used inside nodes nor formulas used for conditional styling are expected to modify the map in any way.
If the formulas ignore it, Freeplane behavior is not determined.

**Reply by macmarrum** [reply]
A very good point.

I should have mentioned that the proposed solution goes outside the contract for scripting in Freeplane, which is read-only access for formulas and Script Filter. The effects might be undesirable, including Freeplane instability.

**Reply by macmarrum** [reply]
The proper approach would be to create a NodeChangeListener with a script. The drawbacks are that it needs to be activated manually each time the mindmap is opened, and that the scripting API is missing a way to keep a reference for the change listener to remove it later, if needed, forcing a user to close/reopen the mindmap to get rid of the listener.

**Reply by euu2021** [reply]
I'm having a hard time understanding this. Can I understand that:

- there exists a "contract for scripting in Freeplane" that states that "neither formulas used inside nodes nor formulas used for conditional styling are expected to modify the map in any way"? 

In other words: is there an actual contract that prohibits any access beyond read-only on the Script Filter? Or I'm getting everything wrong?

**Reply by dpolivaev** [reply]
A contract means Freeplane does not expect you to violate it.
If you try you do it at your own risk.

**Reply by mhechthz** [reply]
Could the option of setting the horizontal shift for a map be a useful feature request?

**Reply by macmarrum** [reply]
It is already available via a script. But from your question I understand that you need something different. Could you please describe these differences or, in general, the requirements?

**Reply by mhechthz** [reply]
No, the script is basically fine, but doesn't permanently change the behaviour of the map - as far as I understood - but correct the spacing once. So it could be a good basic setting for a map.

**Reply by macmarrum** [reply]
I see. So a `Horizontal gap` for all nodes in a map, then?

This got me thinking and I would actually prefer to have it configurable per each node.
There is already `Child gap` property available, which controls the vertical spacing. `Horizontal gap` would control the other direction, but instead of being applied to children, it would impact node's own horizontal spacing, i.e. how far the node is placed from the parent.

Then, to apply it to all nodes, I would set `Horizontal gap` in the `Default` style.

**Reply by EdoFro** [reply]
I like the idea of having an `Horizontal gap`, but I think it should be applied to children, the same way as the `Child Gap` works.

In other words, I would like to have a `Child vertical gap` (existing `Child Gap`) and a `Child horizontal gap`

This way the user can set it to the "**Default style**" and it will act to the whole map automatically

**Comment by macmarrum** [comment]
Nice test cases 👍

Without looking at the source code behind Script Filters and/or Map Conditional Style, I'd have a hard time forming an opinion why there are such differences in behavior.

I've been thinking about a way to auto-execute a script when a minamap opens. I'd like to use an `Attribute` on `root`, named "scriptOnMapOpen", which could contain code to launch NodeChangeListeners. Then to have an init script that adds a ChangeListener of some kind to find and execute `root['scriptOnMapOpen']` when a map is opened. The last part is unclear to me. I'm not sure FP is ready to support that. @dpolivaev any tips?

**Reply by euu2021** [reply]
> I've been thinking about a way to auto-execute a script when a minamap opens. I'd like to use an `Attribute` on `root`, named "scriptOnMapOpen", which could contain code to launch NodeChangeListeners. Then to have an init script that adds a ChangeListener of some kind to find and execute `root['scriptOnMapOpen']` when a map is opened. The last part is unclear to me. I'm not sure FP is ready to support that. @dpolivaev any tips?

This would be enough for people like me, who use only one map, but, for people who use multiple maps, the init scripts are very limited, as we mentioned above:

>> Oh, I see. I guess it worked on a single mindmap that was open automatically during FP start.
> 
> Yes, so it's kind of limited. It's not very useful if the user uses several mindmaps.

---
Maybe, a solution is using a Script Filter for the purpose of starting the listeners. Something like that:

-If map was opened mode than 5 seconds ago: return false
-else: execute the scripts that have the listeners. return true

Just a bad example, but there must be a clever way to write it so it only executes the scripts once, on the startup.

**Reply by macmarrum** [reply]
> This would be enough for people like me, who use only one map, but, for people who use multiple maps, the init scripts are very limited, as we mentioned above

I got to the point where an init script ([scriptOnMapOpen.groovy](https://github.com/macmarrum/freeplane-scripting/blob/main/src/init/scriptOnMapOpen.groovy)) goes over all open mindmaps and executes **scriptOnMapOpen** attribute on the root node. **scriptOnMapOpen**, in turn, could contain instructions to launch a Change Listener.

I am still trying to figure out how to auto-execute **scriptOnMapOpen** when a map is opened after FP is already running. A script can handle opening a mindmap from the UI: a file-chooser dialog would be presented, the selected mindmap opened, and **scriptOnMapOpen** executed. `Ctrl+o` hot key could be assigned to the script. But when a .mm file is double-clicked in the filesystem and passed directly to Freeplane executable, this wouldn't work.

> Maybe, a solution is using a Script Filter for the purpose of starting the listeners. Something like that:
-If map was opened mode than 5 seconds ago: return false
-else: execute the scripts that have the listeners. return true

An idea worth exploring.
The question is how to query from a mindmap the time when it was opened?

**Reply by euu2021** [reply]
> I got to the point where an init script ([scriptOnMapOpen.groovy](https://github.com/macmarrum/freeplane-scripting/blob/main/src/init/scriptOnMapOpen.groovy)) goes over all open mindmaps and executes **scriptOnMapOpen** attribute on the root node. **scriptOnMapOpen**, in turn, could contain instructions to launch a Change Listener.

I will try it.

> An idea worth exploring. The question is how to query from a mindmap the time when it was opened?

Yes, I forgot to consider that this information is not at hand.

Now, what about using the new getListeners() like this on a Script Filter:

- IF getListeners() is not empty: return false
- ELSE:  execute the scripts that have the listeners. And also create a dummy listener, just for the purpose of making the first condition true, which allows deactivating the other listeners if the user wants. return: true

**Comment by dpolivaev** [comment]
I think that any common formatting properties should be always controlled by styles and not configured individually.
Adding the horizontal distance to the parent as a style-controlled property seems to make the most sense and be the most consistent solution.

Do you agree?

**Reply by Illegit** [reply]
I do not know if I'm speaking appropriately, I think that we can reshape the first-level children of a node just by applying a script to that node and be successful to keep the rest of the map untouched.
We can have a script for each shape we want to use
We can think that shape is parametrized with 1 or 2 values.
Maybe if we have a good idea we can have a single script with 1 couple of values (x-axis and y-axis) to create all of the shapes tree, star, org chart, and so on.
In this way we never touch the substance of Freeeplane....

Do you think it feasible? And useful?

**Comment by arresse** [comment]
I'm not sure if this is what you're looking for, but Summary Nodes are what I use to expand the map horizontally.

1) The first node in this configuration acts as your first Topic.

2) Create 3 Child nodes off of this first node.

3) Create a Summary node off of these 3 child nodes.

This is your first Cell.

Then create a second Topic node as a Child off of the Summary node, and repeat steps 2 and 3. This is your second Cell.

You can create as many cells as you need.

Now experiement with adding more Child nodes to the Topic nodes, and also to the Child nodes running off the Topic nodes, and so on. Adjust the Child Gaps as needed.

I've set F1 thru F4 to Fold and unFold nodes. Also Local hyperlinks, and (cloned and uncloned) drop-down Jump Lists via the Note style containing internal and external links to make navigation quick and easy.

I think if you play around with this for a while you'll figure it out. It works very well for me. Just use your imagination . . . Freeplane is an amazing app!
